text/0014-is-authorized-batch.md

```diff
@@ -0,0 +1,247 @@
+# Create a batch version of is_authorized
+
+## Related issues and PRs
+
+- Reference Issues:
+  [k9securityio/cedar-py#13](https://github.com/k9securityio/cedar-py/issues/13)
+- Implementation PR(s): <!-- (leave this empty) -->
+
+## Timeline
+
+- Start Date: 2023-07-11
+- Date Entered FCP: <!-- (leave this empty, update when the PR enters FCP) -->
+- Date Accepted: <!-- (leave this empty, update when the PR is merged) -->
+- Date Landed:
+  <!-- (leave this empty, update when the implementation is in a stable
+  release) -->
+
+## Summary
+
+Initial experiments with language bindings show that while calling
+`is_authorized` for single requests is adequately performant, doing a full
+round-trip many times to satisfy a single higher-level operation incurs
+significant overhead ("consistently 2 orders of magnitude greater \[than the
+cost of calling the function in Rust]").
+
+As a result, we propose a batch version of `is_authorized` that can accept a
+list of authorization requests and provides a matching list of responses,
+eliminating the repeated round-trip overhead.
+
+## Basic examples
+
+### All or nothing
```

Comment:
This use case (composite operations) seems different enough from the others (batch or list APIs) that I think it should probably be a different API.  This one returns a single authz result, while the others are asking for N authz results.

Reply:
@cdisselkoen thanks for the comment. I wrote this example based on my interpretation of a scenario that @skuenzli sketched out for taking multiple requests that was more general than simply multiple resources which I may have misinterpreted. Hoping he can comment with either a better example or more clarity.

Personally I wouldn't necessarily feel a need for a separate core API when the implementation is effectively a `.any` as shown here, but I wouldn't object to one either.

Reply:
I'm interested in an API that accepts:

* multiple requests
* single policy set, schema, and entities

and returns a result for each request.

We populate the context of the request with information relevant to the principal and resource that we write policies against, e.g. `context.principal.name` and `context.resource.name`.

so perhaps something like:

`is_authorized(&self, reqs: Vec<&Request>, p: &PolicySet, e: &Entities) -> Vec<Response>`

It would be the responsibility of the caller to scope the policy set, schema, and entities appropriately to cover the requests.

For my own use cases, I'm interested in batching:
1. multiple actions with a single principal and resource
2. (perhaps) multiple actions and multiple resources with a single principal

The multi-action+multi-resource batching would be beneficial for a common, expensive use case we have, though I haven't done the work to determine the marginal benefit between 1 & 2.  And there are definitely tradeoffs.  In particular, error handling and communication back to the caller may become significantly more complicated if there is a fan-in of hundreds/thousands/more principal-action-resource combinations (true for 1 and more so for 2).  

Everything has limits.

So if Cedar adds a batch-oriented API, I think it's acceptable and wise to enforce some sort of limit, e.g. 1000 p-a-r combos.

Reply:
Thanks @skuenzli!

> We populate the context of the request with information relevant to the principal and resource that we write policies against, e.g. context.principal.name and context.resource.name.

It sounds like it's critical for your use case to have context vary per request, is that correct? There's a [comment](https://github.com/cedar-policy/rfcs/pull/14#discussion_r1261254029) that asks why context isn't constant; seems like this is a reason to keep it as part of each request in the batch.

> So if Cedar adds a batch-oriented API, I think it's acceptable and wise to enforce some sort of limit, e.g. 1000 p-a-r combos.

Could you say more about why you think the batch size limit should be applied at the Cedar level? It seems like the caller could set its own limits on batch size depending on the performance characteristics and risks that it's willing to accept, and Cedar could stay out of it the same way it doesn't put limits on number of entities or policy size.

Reply:
Yes, our context varies by request with both principal and (usually) resource data for that request.

re limits:
Yeah, it might not make sense for the Cedar Authorizer to enforce limits directly.  Perhaps that should be the Cedar user / end service owner's responsibility.

But I'm sure Cedar is designed and built with some constraints and usage patterns in mind.  It's useful to guide users how to use it effectively even if it's just in the library API docs.